opnode/l1/downloader.go

```diff
@@ -0,0 +1,271 @@
+package l1
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru"
+
+	"github.com/ethereum-optimism/optimistic-specs/opnode/eth"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/event"
+)
+
+const (
+	fetchBlockTimeout   = 10 * time.Second
+	fetchReceiptTimeout = 10 * time.Second
+	maxReceiptRetry     = 5
+
+	// 500 at 100 KB each would be 50 MB of memory for the L1 block inputs cache
+	cacheSize = 500
+
+	// Amount of receipt tasks to buffer before applying back-pressure (blocking new block-requests)
+	receiptQueueSize = 100
+)
+
+var (
+	DownloadEvictedErr = errors.New("evicted")
+	DownloadClosedErr  = errors.New("closed")
+)
+
+type downloadTask struct {
+	// Incremented when Finish is called, to check whether the task already completed (with or without error).
+	// May increment >1 when other sub-tasks fail.
+	//
+	// First field, aligned atomic changes.
+	finished uint32
+
+	// Count already downloaded receipts to track status
+	//
+	// Aligned after above field, atomic changes.
+	downloadedReceipts uint32
+
+	block *types.Block
+	// receipts slice is allocated in advance, one slot for each transaction, nil until downloaded
+	receipts []*types.Receipt
+
+	// stop fetching if this context is dead
+	ctx context.Context
+	// to free up resources (e.g. ongoing receipt tasks) if a downloadTask finishes prematurely
+	cancel context.CancelFunc
+
+	// feed to subscribe the requests to (de-duplicate work)
+	feed event.Feed
+}
+
+// wrappedErr wraps an error, since event.Feed cannot handle nil errors otherwise (reflection on nil)
+type wrappedErr struct {
+	error
+}
+
+func (bl *downloadTask) Finish(err wrappedErr) {
+	if atomic.AddUint32(&bl.finished, 1) == 1 {
+		bl.feed.Send(err)
+	}
+}
+
+type receiptTask struct {
+	blockHash common.Hash
+	txHash    common.Hash
+	txIndex   uint64
+	// Count the attempts we made to fetch this receipt. Block as a whole fails if we tried too many times.
+	retry uint64
+	// Avoid concurrent Downloader cache access and pruning edge cases with receipts.
+	// Keep a pointer to the parent task to insert the receipt into.
+	dest *downloadTask
+}
+
+type Downloader interface {
+	Fetch(ctx context.Context, id eth.BlockID) (*types.Block, []*types.Receipt, error)
+	AddReceiptWorkers(n int) int
+	Close()
+}
+
+type DownloadSource interface {
+	eth.BlockByHashSource
+	eth.ReceiptSource
+}
+
+type downloader struct {
+	// cache of ongoing/completed block tasks: block hash -> block
+	cacheEvict *lru.Cache
+	cacheLock  sync.Mutex
+
+	receiptTasks       chan *receiptTask
+	receiptWorkers     []ethereum.Subscription
+	receiptWorkersLock sync.Mutex
+
+	// source to pull data from. Nil when downloader is closed
+	src DownloadSource
+}
+
+func NewDownloader(dlSource DownloadSource) Downloader {
```

Comment:
I tend to prefer to return a struct and let package that uses the downloader define the interface at the point of usage. (https://blog.chewxy.com/2018/03/18/golang-interfaces/)

Reply:
Me too, a struct is generally preferred indeed, but everything except the interface methods is private in this case. And if returning the (`*downloader`), it's easy to make a bad copy by mistake (it's not supposed to be copied because of the mutex and attached workers). Hiding it with the interface in this case protects us from that mistake.

Reply:
I don't really mind that it's only public methods are in the interface. And for the copying, it might be a little more suggestive, but it's still a pointer under the hood and go lets you copy objects that are typed as interfaces..

Reply:
Ofc you can still copy a bare struct if it's typed as interface, but it's `*downloader -> downloader` you can't accidentally do when it's wrapped with the interface.

Reply:
I can see both sides of the argument here. I prefer the current way which is slightly clearer, while less idiomatic. I think what it boils down to is whether there will be other implementations of ` Downloader`. If we expect there will be, it makes sense to return the struct instead, making it clear it's just one possible implementation.

Having `Downloader` defined at the use site would be much less clear in this case imho.